src/Transpile.ts

```diff
@@ -0,0 +1,188 @@
+import * as ts from "typescript";
+import { CompilerOptions } from "./CompilerOptions";
+import { LuaPrinter } from "./LuaPrinter";
+import { LuaTransformer } from "./LuaTransformer";
+import { TranspileError } from "./TranspileError";
+
+function getCustomTransformers(
+    options: CompilerOptions,
+    customTransformers: ts.CustomTransformers,
+    onSourceFile: (sourceFile: ts.SourceFile) => void
+): ts.CustomTransformers {
+    // TODO: https://github.com/Microsoft/TypeScript/issues/28310
+    const forEachSourceFile = (
+        node: ts.SourceFile,
+        callback: (sourceFile: ts.SourceFile) => ts.SourceFile
+    ) =>
+        ts.isBundle(node)
+            ? ((ts.updateBundle(node, node.sourceFiles.map(callback)) as unknown) as ts.SourceFile)
+            : callback(node);
+
+    const luaTransformer: ts.TransformerFactory<ts.SourceFile> = () => node =>
+        forEachSourceFile(node, sourceFile => {
+            onSourceFile(sourceFile);
+            return ts.createSourceFile(sourceFile.fileName, "", ts.ScriptTarget.ESNext);
+        });
+
+    return {
+        afterDeclarations: customTransformers.afterDeclarations,
+        before: [
+            ...(customTransformers.before || []),
+            ...(customTransformers.after || []),
+            luaTransformer,
+        ],
+    };
+}
+
+export interface TranspiledFile {
+    lua?: string;
+    sourceMap?: string;
+    declaration?: string;
+    declarationMap?: string;
+}
+
+export interface TranspilationResult {
+    diagnostics: ts.Diagnostic[];
+    transpiledFiles: Map<string, TranspiledFile>;
+}
+
+export interface GetTranspileOutputOptions {
+    program: ts.Program;
+    options: CompilerOptions;
+    customTransformers?: ts.CustomTransformers;
+    sourceFiles?: ts.SourceFile[];
+    printer?: LuaPrinter;
+    transformer?: LuaTransformer;
+}
+
+export function getTranspileOutput({
```

Comment:
Like I mentioned in the discord, my suggestions for this would be `getTranspilationResult`, `getTranspileResult` or maybe `transpileProgram` (think this one looks too much like the high-level API functions though)

Reply:
What I don't really like about all these names is `get`, to me such name implies a somewhat simple pure function, not a heavy action. What do you think about just `transpile`? It's still close to high-level API, but a bit less than `transpileProgram`.

Reply:
I was also thinking just `transpile` would be a decent option. I'm fine with switching to that.